Lis2dh12 driver

DanielLockau-MLPA opened the pull request:
For discussion of the source branch only! We don't want to merge this!

@benpicco Please browse this draft PR and comment if you have any quick suggestions for this. I generally see it as an improvement worth a PR into RIOT but it still needs to be cleaned out a bit.

Commit:
changed lis2dh12 driver for interrupt use

Commit:
tested interrupt function for setting and reading

removed comments

added description for function

DanielLockau-MLPA commented on drivers/include/lis2dh12.h:
```diff
@@ -116,11 +116,53 @@ enum {
+int lis2dh12_set_interrupt(const lis2dh12_t *dev, int_params_t params, uint8_t INT_X);
+
+/**
+ * @brief   read an interruptevent on LIS2DH12 sensor device
```
Please capitalize first letter of the brief description to be in line with the rest of this file

DanielLockau-MLPA replied:
Also please separate "interruptevent" into "interrupt event".

DanielLockau-MLPA commented on drivers/include/lis2dh12.h:
```diff
@@ -116,11 +116,53 @@ enum {
 extern const saul_driver_t lis2dh12_saul_driver;
 
+/**
+ * @brief   Set the Interruptvalues in LIS2DH12 sensor device
```
Please only use lower case words (apart from first word names like LIS2DH12) in descriptions. Also, please separate "Interruptvalues" into "interrupt values".

DanielLockau-MLPA commented on drivers/include/lis2dh12.h:
```diff
@@ -116,11 +116,53 @@ enum {
 };
 
+/**
+ * @brief    interrupt section
```
This struct needs more and possibly per-field comments to be usable.

DanielLockau-MLPA commented on drivers/include/lis2dh12.h:
```diff
@@ -116,11 +116,60 @@ enum {
+ * @return  LIS2DH12_OK on success
+ * @return  LIS2DH12_NOBUS on bus errors
+ */
+int lis2dh12_write(const lis2dh12_t *dev, uint8_t reg, uint8_t data);
```
I would suggest to rename this to `lis2dh12_write_reg`. My last comment was on an old commit, I'll delete it.

DanielLockau-MLPA replied:
Could you write some words to the discussion to specify why you want to expose the register write function to the user?

jan-mo replied:
this is only for the SAUL .write function, it is not necessary. I will remove it.

DanielLockau-MLPA commented on drivers/include/lis2dh12.h:
```diff
@@ -116,11 +116,60 @@ enum {
+
+/**
+ * @brief   read an interruptevent on LIS2DH12 sensor device
+ *
```
You should add a longer description here about what happens when executing the function (what kind of data will be returned in `*data`). This is not clear from the brief description. Maybe you can also find a more concise wording for the brief.

DanielLockau-MLPA commented on drivers/include/lis2dh12.h:
```diff
@@ -116,11 +116,60 @@ enum {
+int lis2dh12_read_interrupt(const lis2dh12_t *dev, uint8_t *data, uint8_t INT_X);
+
+/**
+ * @brief write Data to specific Register
```
Please capitalize first letter of the first word and lower case all other words.

DanielLockau-MLPA commented on drivers/include/lis2dh12.h:
```diff
@@ -116,11 +116,60 @@ enum {
+ * @brief write Data to specific Register
+ *
+ * @param[in] dev   device descriptor
+ * @param[in] reg   specific Register, number must be decimal
```
all lower case please

DanielLockau-MLPA commented on drivers/include/lis2dh12.h:
```diff
@@ -116,11 +116,60 @@ enum {
+ * @return  LIS2DH12_OK on success
+ * @return  LIS2DH12_NOBUS on bus errors
+ */
+int lis2dh12_read_interrupt(const lis2dh12_t *dev, uint8_t *data, uint8_t INT_X);
```
Please do not use capital variable names like `INT_X` here. The name `INT_X` is also not very descriptive. I would suggest something like `interrupt_line`.

DanielLockau-MLPA replied:
I would also suggest abbreviating the method name to `lis2dh12_read_int_src` to be more precise on what the user gets.

DanielLockau-MLPA commented on drivers/include/lis2dh12.h:
```diff
@@ -116,11 +116,60 @@ enum {
+ *
+ * @param[in] dev      device descriptor
+ * @param[out] data    device interrupt data
+ * @param[in] INT_X    number of interrupt line 
```
I would make the description more verbose. `number of the interrupt line (1 or 2)`

DanielLockau-MLPA commented on drivers/include/lis2dh12.h:
```diff
@@ -116,11 +116,60 @@ enum {
+    uint8_t type;
+} int_params_t;
+
+#define INT_1 0x1
```
I would leave out these trivial defines and rather include the possible values in the documentation. Their names also lack prefixing and therefore have a higher probability of colliding with definitions in other headers or compilation routines.

DanielLockau-MLPA commented on drivers/include/lis2dh12.h:
```diff
@@ -116,11 +116,60 @@ enum {
+    uint8_t ths;
+    uint8_t duration;
+    uint8_t type;
+} int_params_t;
```
I suggest prefixing this type name.

DanielLockau-MLPA replied:
1. I am not sure you got this right: I was referring to the type name `int_params_t` -> `lis2dh12_int_params_t`, not the field names.

DanielLockau-MLPA commented on drivers/include/lis2dh12.h:
```diff
@@ -116,11 +116,60 @@ enum {
+ */
+typedef struct {
+    uint8_t cfg;
+    uint8_t ths;
```
Please don't unnecessarily abbreviate names. This probably means `threshold`.

DanielLockau-MLPA replied:
(there are some conventional abbreviations but I would avoid using too many abbreviations in the interface)

DanielLockau-MLPA commented on drivers/lis2dh12/lis2dh12.c:
```diff
@@ -187,6 +187,55 @@ int lis2dh12_read(const lis2dh12_t *dev, int16_t *data)
+    assert(INT_X == INT_1 || INT_X == INT_2);
+
+    _acquire(dev);
+    if (INT_X == INT_1){
```
Please replace `if...else if...` with a `switch` and also throw an error in the `default` case.

DanielLockau-MLPA replied:
(don't forget to `_release` in case of error)

jan-mo replied:
I removed `assert(INT_X == 1 ...)`, because it is handled in switch default.

DanielLockau-MLPA commented on drivers/include/lis2dh12.h:
```diff
@@ -116,11 +116,60 @@ enum {
+ * @brief    interrupt section
+ */
+typedef struct {
+    uint8_t cfg;
```
Please turn this configuration into a bitmask-type struct. There is little added value over the direct register write if the user has to pass an oblique `uint8_t` configuration type. Suggested structure:

```C
typedef struct {
    union {
        struct {
            unsigned field_1 :1;
            unsigned field_2 :1;
            ...
        } bits;
        uint8_t byte;
    } config;
    ...
}
```

Using this structure the user can set the bits and you can read the byte when sending the configuration.

jan-mo replied:
not done yet

DanielLockau-MLPA commented on drivers/lis2dh12/lis2dh12.c:
```diff
@@ -187,6 +187,55 @@ int lis2dh12_read(const lis2dh12_t *dev, int16_t *data)
 
+int lis2dh12_write(const lis2dh12_t *dev, uint8_t reg, uint8_t data)
+{
+    _write(dev, reg, data);
```
~~As this is a trivial wrapper around a static function I would suggest to change the signature of the static function.~~

EDIT: You might want to return the i2c exit status from _write and transform it to a lis2dh12 exit value. Also, you are forgetting to _acquire() and _release() the bus. Please make sure this is done properly if you decide to leave this function in.

DanielLockau-MLPA commented on drivers/lis2dh12/lis2dh12.c:
```diff
@@ -187,6 +187,55 @@ int lis2dh12_read(const lis2dh12_t *dev, int16_t *data)
+}
+
+/* read interrupt INT_X (INT_1 or INT_2) */
+int lis2dh12_read_interrupt(const lis2dh12_t *dev, uint8_t *data, uint8_t INT_X)
```
Reading the code suggests that the function is not ideally named. I would suggest `lis2dh12_read_int_src(dev, src, interrupt_line)` as signature. I would also suggest to make `src` a less opaque type. You could make it an enum or bitfield, perferably an enum (be sure to include a `LIS2DH12_INT_SRC_NONE` (or similar name) default value in the enum).

DanielLockau-MLPA commented on drivers/include/lis2dh12.h:
```diff
@@ -116,11 +116,60 @@ enum {
+ * @return  LIS2DH12_OK on success
+ * @return  LIS2DH12_NOBUS on bus errors
+ */
+int lis2dh12_set_interrupt(const lis2dh12_t *dev, int_params_t params, uint8_t INT_X);
```
I would suggest abbreviating to `lis2dh12_set_int` because this abbreviation is commonly used in function names in RIOT.

DanielLockau-MLPA commented on drivers/lis2dh12/lis2dh12.c:
```diff
@@ -187,6 +187,55 @@ int lis2dh12_read(const lis2dh12_t *dev, int16_t *data)
+    assert(INT_X == INT_1 || INT_X == INT_2);
+
+    _acquire(dev);
+    if(INT_X == 1)
```
Please put this into a switch and handle the default case with error.

DanielLockau-MLPA commented on drivers/lis2dh12/lis2dh12_saul.c:
```diff
@@ -31,8 +31,19 @@ static int read_accelerometer(const void *dev, phydat_t *res)
 
+static int write_accelerometer(const void *dev, phydat_t *data)
+{
+    printf("Reg: 0x%02X\n",data->val[0]);
```
Maybe I am missing some insight here: What is the purpose of these printf's? Is this function only for command line usage and this will be part of the output? If this is still for your debug purposes, please remove it.

DanielLockau-MLPA replied:
(I thought saul was also for non-console usage, but I never did actually use it.)

DanielLockau-MLPA commented on drivers/lis2dh12/lis2dh12_saul.c:
```diff
@@ -31,8 +31,19 @@ static int read_accelerometer(const void *dev, phydat_t *res)
+    if (lis2dh12_write((const lis2dh12_t *)dev, data->val[0], data->val[1]) != LIS2DH12_OK) {
+        return 0;
+    }
+    return 3;
```
Are there any defines for these integer return values?

DanielLockau-MLPA commented on examples/saul/Makefile:
```diff
@@ -9,6 +9,9 @@ RIOTBASE ?= $(CURDIR)/../..
 # we want to use SAUL:
 USEMODULE += saul_default
+
+USEMODULE += lis2dh12
```
I don't think upstream RIOT administrators will allow you to add the LIS2DH12 as dependency to the general saul example. Please remove these changes.

DanielLockau-MLPA commented on examples/saul/main.c:
```diff
@@ -19,14 +19,56 @@
  */
 
 #include <stdio.h>
+#include <string.h>
```
I don't think upstream RIOT administrators will allow you to add the LIS2DH12 as dependency to the general saul example. Please remove these changes.

benpicco replied:
`tests/driver_lis2dh12` would be a better place for this.

DanielLockau-MLPA commented on drivers/lis2dh12/lis2dh12.c:
```diff
@@ -187,6 +187,55 @@ int lis2dh12_read(const lis2dh12_t *dev, int16_t *data)
+/* read interrupt INT_X (INT_1 or INT_2) */
+int lis2dh12_set_interrupt(const lis2dh12_t *dev, int_params_t params, uint8_t INT_X)
+{
+    assert(dev && params.cfg && params.ths && params.duration && params.type);
```
Would you need additional limit checking of the threshold and duration parameters or is the entire uint8_t range allowed? If there are tighter limits, please also document them.

DanielLockau-MLPA replied:
Is it required in all cases that threshold and duration are not equal to zero? This should also be included in the parameter documentation.

jan-mo replied:
7 bits for duration and threshold, zero is allowed. Added this.

DanielLockau-MLPA reviewed (commented):
I think we can generally open a PR on upstream RIOT from this branch once the user interface has been made a little more comfortable (knowledge about the register values has been put into data structures) and the other small issues have been addressed.

Commit:
changed code for PR 2

Commit:
added test for lis2dh12 driver

DanielLockau-MLPA commented on examples/saul/Makefile:
```diff
@@ -9,6 +9,7 @@ RIOTBASE ?= $(CURDIR)/../..
 
 # we want to use SAUL:
 USEMODULE += saul_default
+
```
Please checkout this file from master. It needs to be completely reset.

DanielLockau-MLPA commented on tests/driver_lis2dh12/Makefile:
```diff
@@ -1,7 +1,7 @@
-# as default we use the SPI mode for now, as the I2C mode is not supported, yet
-DRIVER ?= lis2dh12_spi
+# use lis2dh12_spi for SPI-Mode and lis2dh12 for I2C-Mode
+DRIVER ?= lis2dh12
```
I would leave the SPI default.

Shouldn't the other variant be called `lis2dh12_i2c`? Or doesn't the comment make sense any more?

jan-mo replied:
changed to lis2dh12_i2c 

see [https://github.com/RIOT-OS/RIOT/pull/12290](https://github.com/RIOT-OS/RIOT/pull/12290)

DanielLockau-MLPA commented on tests/driver_lis2dh12/main.c:
```diff
@@ -48,6 +73,23 @@ int main(void)
 
+    /* enable interrupt Pins */
+    char* pin = "PA12";
+    if(gpio_init_int(GPIO_PIN(PA,12),GPIO_IN, GPIO_RISING,int_cb,pin) == -1)
```
Compiling this code makes the assumption that the lis2dh12 interrupts are wired to the GPIO_PIN(PA, 12) and GPIO_PIN(PA, 13).

This will not pass the review for a merge into RIOT.

You could either:
- add the interrupt pins as optional module configuration for the lis2dh12 driver, but then you will have to do make the driver configurable with the interrupt callback and handle the interrupts internally - this would require larger modifications
- write a note about mandatory defines for running this test and don't test with interrupts if the pins are not defined

@benpicco Any opinion about this?

jan-mo replied:
I added a TODO firstly. Can be changed later...

Commit:
Driver is now lis2dh12_i2c or lis2dh12_spi

changed IRQ-pins setting, pins must be set by hand

DanielLockau-MLPA commented on tests/driver_lis2dh12/main.c:
```diff
@@ -35,6 +38,33 @@ static char str_out[3][8];
 static lis2dh12_t dev;
 
+/* interrupt pins*/
+//TODO: change the interruptpins to real connected
```
Definition of interrupt pins and enabling / disabling code sections should be done using macros / defines.

Please remove the character constants below and replace them with defines. Enable and disable the sections for the interrupt tests accordingly: exclude all if no pin is defined, include callback and relevant sections for defined interrupt pins.

jan-mo replied:
I added `LIS2DH12_INTERRUPT_TEST` to `main.c`. Maybe there is a better place...

DanielLockau-MLPA commented on drivers/include/lis2dh12.h:
```diff
@@ -114,13 +114,61 @@ enum {
+    uint8_t LIS2DH12_INT_SRC_ZL:1;  /**<  */
+    uint8_t LIS2DH12_INT_SRC_ZH:1;  /**<  */
+    uint8_t LIS2DH12_INT_SRC_IA:1;  /**<  */
+} int_src_reg_t;
```
As for `int_params_t` I would suggest prefixing this name with the module name, e.g. `lis2dh12_int_params_t`.

Commit:
Added LIS2DH12_INTERRUPT_TEST for testing with or w/o irq

Added comments for INT_SRC register and updated struct names

Commit:
Changed define for interrupt usage

Moved LIS2DH12_INT_PIN_1 and LIS2DH12_INT_PIN_2 to board.h

Commit:
Changed LIS2DH12_INT_PIN_1 to LIS2DH12_INT_PIN1 (also for pin 2). Was already set in board.h

DanielLockau-MLPA commented on tests/driver_lis2dh12/main.c:
```diff
@@ -48,6 +79,25 @@ int main(void)
+    lis2dh12_int_params_t params = {0};
+    params.int_type = 0b01000000;
+    params.int_config = 1;
+    params.int_threshold = 0b00011111;
```
This section should be guarded and executed only if interrupt lines are defined.

Commit:
Guarded the params for first and second interrupt

DanielLockau-MLPA commented on tests/driver_lis2dh12/main.c:
```diff
@@ -48,6 +79,32 @@ int main(void)
+    lis2dh12_int_params_t params_int2 = {0};
+    params_int2.int_type = 0b01000000;
+    params_int2.int_config = 1;
+    params_int2.int_threshold = 0b00011111;
```
Use to actual number here for clarity.

DanielLockau-MLPA commented on tests/driver_lis2dh12/main.c:
```diff
@@ -48,6 +79,32 @@ int main(void)
+    lis2dh12_int_params_t params_int1 = {0};
+    params_int1.int_type = 0b01000000;
+    params_int1.int_config = 1;
+    params_int1.int_threshold = 0b00011111;
```
Use to actual number here for clarity.

DanielLockau-MLPA commented on tests/driver_lis2dh12/main.c:
```diff
@@ -48,6 +79,32 @@ int main(void)
+
+    /* create and set the interrupt params */
+    lis2dh12_int_params_t params_int1 = {0};
+    params_int1.int_type = 0b01000000;
```
Use bitwise or of defined constants to set int type.

DanielLockau-MLPA commented on tests/driver_lis2dh12/main.c:
```diff
@@ -48,6 +79,32 @@ int main(void)
+
+    /* create and set the interrupt params */
+    lis2dh12_int_params_t params_int2 = {0};
+    params_int2.int_type = 0b01000000;
```
Use bitwise or of defined constants to set int type.

Commit:
Added interrupt types and config settings

Changed values to human-readable

benpicco commented on drivers/lis2dh12/lis2dh12.c:
```diff
@@ -187,6 +187,81 @@ int lis2dh12_read(const lis2dh12_t *dev, int16_t *data)
+    _acquire(dev);
+
+    uint8_t buffer = 0;
+    (void)buffer;
```
This is not needed.

benpicco commented on drivers/lis2dh12/lis2dh12.c:
```diff
@@ -187,6 +187,81 @@ int lis2dh12_read(const lis2dh12_t *dev, int16_t *data)
+        default:
+            _release(dev);
+            return LIS2DH12_NOINT;
+            break;
```
No need to `break` after `return`.

benpicco commented on drivers/include/lis2dh12.h:
```diff
@@ -114,13 +114,96 @@ enum {
+    uint8_t LIS2DH12_INT_SRC_ZL:1;  /**< Z low event has occured */
+    uint8_t LIS2DH12_INT_SRC_ZH:1;  /**< Z high event has occured */
+    uint8_t LIS2DH12_INT_SRC_IA:1;  /**< 1 if interrupt occured */
+} lis2dh12_int_src_reg_t;
```
Seems a bit over-complicated to turn this into a bit-field.
The more common

```C
#define LIS2DH12_INT_SRC_XL 0x01
#define LIS2DH12_INT_SRC_XH 0x02
#define LIS2DH12_INT_SRC_YL 0x04
#define LIS2DH12_INT_SRC_YH 0x08
#define LIS2DH12_INT_SRC_ZL 0x10
#define LIS2DH12_INT_SRC_ZH 0x20
#define LIS2DH12_INT_SRC_IA 0x40
```

should suffice.
Then you don't have to copy individual bits and can still easily check the individual events (`data & LIS2DH12_INT_SRC_XL`)

DanielLockau-MLPA replied:
Please document the source register retrieval method's interface accordingly if you decide to use defines here.

jan-mo replied:
created a union around struct

benpicco commented on drivers/lis2dh12/lis2dh12.c:
```diff
@@ -187,6 +187,81 @@ int lis2dh12_read(const lis2dh12_t *dev, int16_t *data)
+        default:
+            _release(dev);
+            return LIS2DH12_NOINT;
+            break;
```
No need for the `break`.

benpicco commented on drivers/lis2dh12/lis2dh12.c:
```diff
@@ -187,6 +187,81 @@ int lis2dh12_read(const lis2dh12_t *dev, int16_t *data)
+    data->LIS2DH12_INT_SRC_YH = buffer >> 3;
+    data->LIS2DH12_INT_SRC_ZL = buffer >> 4;
+    data->LIS2DH12_INT_SRC_ZH = buffer >> 5;
+    data->LIS2DH12_INT_SRC_IA = buffer >> 6;
```
No need for that if you drop the bit field.

benpicco commented on tests/driver_lis2dh12/main.c:
```diff
@@ -35,6 +38,34 @@ static char str_out[3][8];
+        return;
+    }
+
+    printf("content SRC_Reg:\n\t XL 0x%02x\n",buffer.LIS2DH12_INT_SRC_XL);
```
```suggestion
    printf("content SRC_Reg:\n\t XL 0x%02x\n",!!(buffer & LIS2DH12_INT_SRC_XL));
```
and so on

benpicco commented on tests/driver_lis2dh12/main.c:
```diff
@@ -35,6 +38,34 @@ static char str_out[3][8];
+
+    lis2dh12_int_src_reg_t buffer = {0};
+
+    if(!strcmp("INT_1",pin)){
```
```suggestion
    if (!strcmp("INT_1",pin)) {
```

benpicco commented on drivers/include/lis2dh12.h:
```diff
@@ -114,13 +114,96 @@ enum {
+    LIS2DH12_INT_1_TYPE_CLICK   = 0x80, /**< click interrupt on INT_1 */
+
+    /* for interrupt 2 (CTRL_REG6) */
+    LIS2DH12_INT_2_TYPE_POLARITY = 0x02, /**< INT_1 and INT_2 pin polarity (0: high-active; 1: low-active) */
```
You can't have two elements with the same value in an `enum`.
Just use `#define`.

DanielLockau-MLPA replied:
Wow! How did the compiler not warn about that?

benpicco replied:
Turns out this is actually legal in C.

Commit:
Added some changes

Commit:
Created union around INT_SRC register

benpicco commented on drivers/include/lis2dh12.h:
```diff
@@ -114,13 +114,100 @@ enum {
+        uint8_t LIS2DH12_INT_SRC_ZH:1;  /**< Z high event has occured */
+        uint8_t LIS2DH12_INT_SRC_IA:1;  /**< 1 if interrupt occured */
+    };
+    uint8_t LIS2DH12_INT_SRC;
```
please use lowercase for variables.
Also no need to prefix struct members with `LIS2DH12`.

Commit:
Changed name of INT_SRC register entry

benpicco commented:
<details>
  <summary>After receiving the Interrupt, the test stops reading values:</summary>

```
2020-01-29 15:25:09,834 # X:    0.019 Y:   -0.023 Z:    1.054
2020-01-29 15:25:09,933 # X:    0.027 Y:    0.039 Z:    0.992
2020-01-29 15:25:10,034 # X:    0.050 Y:    0.019 Z:    1.007
2020-01-29 15:25:10,135 # X:    0.000 Y:   -0.035 Z:    1.109
2020-01-29 15:25:10,234 # X:    0.023 Y:   -0.023 Z:    1.046
2020-01-29 15:25:10,335 # X:    0.042 Y:   -0.035 Z:    1.070
2020-01-29 15:25:10,435 # X:    0.156 Y:   -0.046 Z:    1.078
2020-01-29 15:25:10,535 # X:    0.187 Y:   -0.250 Z:    0.871
2020-01-29 15:25:10,639 # X:    0.519 Y:    0.375 Z:    0.531
2020-01-29 15:25:10,650 # interrupt received from INT_1
2020-01-29 15:25:10,652 # content SRC_Reg:
2020-01-29 15:25:10,652 # 	 XL 0x01
2020-01-29 15:25:10,654 # 	 XH 0x00
2020-01-29 15:25:10,654 # 	 YL 0x01
2020-01-29 15:25:10,655 # 	 YH 0x00
2020-01-29 15:25:10,658 # 	 ZL 0x00
2020-01-29 15:25:10,658 # 	 ZH 0x01
2020-01-29 15:25:10,659 # 	 IA 0x01
2020-01-29 15:25:10,659 # 
2020-01-29 15:25:10,736 # X:    0.945 Y:   -0.238 Z:    1.343
2020-01-29 15:25:10,820 # interrupt received from INT_1
2020-01-29 15:25:10,823 # content SRC_Reg:
2020-01-29 15:25:10,823 # 	 XL 0x01
2020-01-29 15:25:10,825 # 	 XH 0x00
2020-01-29 15:25:10,826 # 	 YL 0x01
2020-01-29 15:25:10,826 # 	 YH 0x00
2020-01-29 15:25:10,829 # 	 ZL 0x00
2020-01-29 15:25:10,829 # 	 ZH 0x01
2020-01-29 15:25:10,830 # 	 IA 0x01
2020-01-29 15:25:10,830 # 
2020-01-29 15:25:10,836 # X:    0.390 Y:   -0.140 Z:    0.683
2020-01-29 15:25:10,936 # X:    1.121 Y:    0.355 Z:    0.464
2020-01-29 15:25:11,038 # X:    0.664 Y:   -0.234 Z:    0.570
2020-01-29 15:25:11,137 # X:    0.621 Y:    0.058 Z:    0.781
2020-01-29 15:25:11,236 # interrupt received from INT_1
2020-01-29 15:25:11,237 # content SRC_Reg:
2020-01-29 15:25:11,238 # 	 XL 0x01
2020-01-29 15:25:11,238 # 	 XH 0x00
2020-01-29 15:25:11,239 # 	 YL 0x01
2020-01-29 15:25:11,239 # 	 YH 0x00
2020-01-29 15:25:11,241 # 	 ZL 0x00
2020-01-29 15:25:11,242 # 	 ZH 0x01
2020-01-29 15:25:11,243 # 	 IA 0x01
2020-01-29 15:25:11,243 # 
2020-01-29 15:25:11,246 # X:    0.136 Y:    0.140 Z:    1.117
2020-01-29 15:25:11,293 # interrupt received from INT_1
2020-01-29 15:25:11,296 # content SRC_Reg:
2020-01-29 15:25:11,296 # 	 XL 0x01
2020-01-29 15:25:11,305 # 	 XH 0x00
2020-01-29 15:25:11,306 # 	 YL 0x01
2020-01-29 15:25:11,306 # 	 YH 0x00
2020-01-29 15:25:11,307 # 	 ZL 0x00
2020-01-29 15:25:11,307 # 	 ZH 0x01
2020-01-29 15:25:11,307 # 	 IA 0x01
2020-01-29 15:25:11,308 # 
2020-01-29 15:25:11,337 # X:    0.218 Y:   -0.007 Z:    0.980
2020-01-29 15:25:11,438 # X:    0.218 Y:    0.000 Z:    1.011
2020-01-29 15:25:11,538 # X:    0.082 Y:   -0.003 Z:    0.812
2020-01-29 15:25:11,638 # X:    0.257 Y:   -0.089 Z:    1.164
2020-01-29 15:25:11,736 # interrupt received from INT_1
2020-01-29 15:25:11,739 # content SRC_Reg:
2020-01-29 15:25:11,739 # 	 XL 0x00
2020-01-29 15:25:11,741 # 	 XH 0x00
2020-01-29 15:25:11,742 # 	 YL 0x00
2020-01-29 15:25:11,742 # 	 YH 0x00
2020-01-29 15:25:11,744 # 	 ZL 0x00
2020-01-29 15:25:11,744 # 	 ZH 0x00
2020-01-29 15:25:11,745 # 	 IA 0x00
2020-01-29 15:25:11,746 # 
2020-01-29 15:25:11,749 # X:    0.390 Y:   -0.250 Z:    0.000
2020-01-29 15:25:11,778 # interrupt received from INT_1
2020-01-29 15:25:11,802 # content SRC_Reg:
2020-01-29 15:25:11,802 # 	 XL 0x00
2020-01-29 15:25:11,805 # 	 XH 0x00
2020-01-29 15:25:11,805 # 	 YL 0x00
2020-01-29 15:25:11,805 # 	 YH 0x00
2020-01-29 15:25:11,809 # 	 ZL 0x00
2020-01-29 15:25:11,809 # 	 ZH 0x00
2020-01-29 15:25:11,809 # 	 IA 0x00
2020-01-29 15:25:11,810 # 
2020-01-29 15:25:11,861 # X:    0.000 Y:    0.000 Z:    0.000
2020-01-29 15:25:11,879 # interrupt received from INT_1
2020-01-29 15:25:11,903 # content SRC_Reg:
2020-01-29 15:25:11,904 # 	 XL 0x00
2020-01-29 15:25:11,905 # 	 XH 0x00
2020-01-29 15:25:11,906 # 	 YL 0x00
2020-01-29 15:25:11,906 # 	 YH 0x00
2020-01-29 15:25:11,909 # 	 ZL 0x00
2020-01-29 15:25:11,910 # 	 ZH 0x00
2020-01-29 15:25:11,910 # 	 IA 0x00
2020-01-29 15:25:11,910 # 
2020-01-29 15:25:11,920 # interrupt received from INT_1
2020-01-29 15:25:11,943 # content SRC_Reg:
2020-01-29 15:25:11,944 # 	 XL 0x00
2020-01-29 15:25:11,946 # 	 XH 0x00
2020-01-29 15:25:11,947 # 	 YL 0x00
2020-01-29 15:25:11,948 # 	 YH 0x00
2020-01-29 15:25:11,950 # 	 ZL 0x00
2020-01-29 15:25:11,951 # 	 ZH 0x00
2020-01-29 15:25:11,951 # 	 IA 0x00
2020-01-29 15:25:11,952 # 
2020-01-29 15:25:11,976 # X:    0.000 Y:    0.000 Z:    0.000
2020-01-29 15:25:12,061 # X:    0.000 Y:    0.000 Z:    0.000
2020-01-29 15:25:12,162 # X:    0.000 Y:    0.000 Z:    0.000
2020-01-29 15:25:12,262 # X:    0.000 Y:    0.000 Z:    0.000
2020-01-29 15:25:12,362 # X:    0.000 Y:    0.000 Z:    0.000
```
</details>

jan-mo commented:
> After receiving the Interrupt, the test stops reading values:

I don´t have this issue. Maybe the lis2dh12 did not loaded the right configs.
The lis2dh12 only reload config registers after reboot. The supply voltage needs to be removed after flashing the board.

benpicco commented:
> The lis2dh12 only reload config registers after reboot.

Then we should probably make use of the Reset pin to ensure the right values are used.

jan-mo commented:
> Then we should probably make use of the Reset pin to ensure the right values are used.

I think the lis device gets to many requests and run in a blocking state or so.
I put a very short delay between the readings and after some interrupts i get the zeros. After restart device by flashing, the lis can not be initialize.

So I think it is not an issue.

```
2020-02-07 10:15:32,320 # X:    0.000 Y:    0.000 Z:    0.000
2020-02-07 10:15:32,421 # X:    0.000 Y:    0.000 Z:    0.000
2020-02-07 10:15:38,823 # main(): This is RIOT! (Version: 2020.01-devel-830-g268a3-lis2dh12_driver)
2020-02-07 10:15:38,827 # LIS2DH12 accelerometer driver test application
2020-02-07 10:15:38,828 # 
2020-02-07 10:15:38,830 # Initializing LIS2DH12 sensor... 
2020-02-07 10:15:38,863 # [lis2dh12] error: invalid value read from WHO_AM_I register
2020-02-07 10:15:38,865 # [Failed]
2020-02-07 10:15:38,865 # 
```

Commit:
Added function to check STATUS_REG on device

Changed read of content from interrupt register

Added mutex and flags for interrupt handling

benpicco commented on tests/driver_lis2dh12/main.c:
```diff
@@ -48,25 +93,69 @@ int main(void)
-    xtimer_ticks32_t last_wakeup = xtimer_now();
+    /* enable interrupt Pins */
+#ifdef LIS2DH12_INT_PIN1
+    if (gpio_init_int(LIS2DH12_INT_PIN1,GPIO_IN, GPIO_RISING,lis2dh12_int_cb,"INT_1") == -1)
```
I'd move that into `lis2dh12_set_int()`.
Also don't hard-code `LIS2DH12_INT_PIN1` but rather add an entry to `lis2dh12_params_t` - then you don't have to specify the pin but can just use the `int_line` parameter.

It also makes it possible to have more than one instance of the lis2dh12 driver running, we should not regress this if we want to get this merged.

benpicco commented on tests/driver_lis2dh12/main.c:
```diff
@@ -35,6 +39,47 @@ static char str_out[3][8];
+/* timer lock */
+mutex_t lock_int = MUTEX_INIT;
+bool int_1_occ = false;
+bool int_2_occ = false;
```
```C
struct lis_ctx {
    mutex_t *lock;
    uint8_t *flags;
    uint8_t line;
};

static void lis2dh12_int_cb(void* _ctx) {
    struct lis_ctx *ctx = _ctx;

    *ctx->flags |= (1 << ctx->line);
    mutex_unlock(ctx->lock);
}
```

benpicco commented on tests/driver_lis2dh12/main.c:
```diff
@@ -48,25 +93,69 @@ int main(void)
         return 1;
     }
 
```
```C
mutex_t lock_int = MUTEX_INIT_LOCKED;
uint8_t flags = 0;

struct lis_ctx ctx[2] = {
    {
        .lock = &mutex
        .flags = &flags;
        .line = 0;
    }, {
        .lock = &mutex
        .flags = &flags;
        .line = 1;
    }
} ;
```

benpicco commented on tests/driver_lis2dh12/main.c:
```diff
@@ -48,25 +93,69 @@ int main(void)
+    params_int1.int_config = LIS2DH12_INT_CFG_XLIE;
+    params_int1.int_threshold = 31;
+    params_int1.int_duration = 1;
+    lis2dh12_set_int(&dev,params_int1,1);
```
```suggestion
    lis2dh12_set_int(&dev, params_int1, 1, &ctx[0]);
```

Commit:
Added interrupt handling with messages

Found bug in xtimer_msg_receive_timeout and xtimer_mutex_lock_timeout

benpicco committed:
tests/driver_lis2dh12: clean up testcase, use mutex

I cleaned up the testcase a bit and changed it to use a mutex.
It doesn't crash anymore now.

benpicco commented on drivers/include/lis2dh12.h:
```diff
@@ -144,6 +248,17 @@ int lis2dh12_init(lis2dh12_t *dev, const lis2dh12_params_t *params);
+ * @return  LIS2DH12_OK on success
+ * @return  LIS2DH12_NOBUS on bus error
+ */
+int lis2dh12_read_status_reg(const lis2dh12_t *dev, lis2dh12_status_reg_t *data);
```
Why is this needed?

benpicco committed:
drivers/lis2dh12: simplify the driver a bit

Use common coding standards, don't define special register types.

benpicco committed:
drivers/lis2dh12: configure interrupt pin inside lis2dh12_set_int()

benpicco committed:
drivers/lis2dh12: don't copy params struct in lis2dh12_set_int()

benpicco commented:
I've reworked the driver a bit.
For consistency I renamed `LIS2DH12_INT_PIN*`-> `LIS2DH12_PARAM_INT_PIN*`, so this will require 
https://ml-pa.visualstudio.com/Liebherr/_git/LBC_CraneAsset_EmbeddedRiotTrackerLargeBoardV2/pullrequest/1017

jan-mo committed:
Added periph_gpio_irq to Makefile.dep

jan-mo commented:
merged in RIOT/master
